### ISSUE TITLE
Source generator output files list is unreliable/bugged

### ISSUE THREAD

**Issue by Sergio0694** [dotnet/roslyn#51810]
**Version Used**: VS 16.10 Preview 1

**Steps to Reproduce**:

1. Clone [ComputeSharp](https://github.com/Sergio0694/ComputeSharp)
2. Open the solution, build a test project (for instance `ComputeSharp.Tests.Internals`)
3. Inspect the analyzer outputs in Solution Explorer, then do any of the following:
    - Make some changes to a unit test (eg. delete some)
    - Close the solution, delete `\bin` and `\obj`, reopen and rebuild
4. The SG output files will sometimes not be displayed at all, or not update after some changes

Note that the solution still builds just fine, and the generated code actually works fine.
It just is not visible/updated in the Solution Explorer preview.

Additionally, when eg. the preview files are not displayed at all, VS also shows red squiggly lines everywhere, and the generated files are displayed in white (no proper IntelliSense/highlight, etc.). Despite this though, the project still builds and runs just fine.

**Expected Behavior**:
Making any changes and rebuilding should correctly update the preview files.

**Actual Behavior**:
The preview files are unreliable, sometimes they're not updated or generated at all.

**Screenshot**:
![image](https://user-images.githubusercontent.com/10199417/110855164-4cdad280-82b6-11eb-9feb-6eb3a0a796ec.png)

**Comment by maikschott** [comment]
I have found (VS 16.9.4.) that the generator output files list is only shown when the generator is used in a root project (i.e. a console project), but not if the generator is used in a library project. The same with Sergio's other issues.

**Comment by sharwell** [comment]
This is generally expected. When your SG is defined in the same solution where it is used, Visual Studio will use the source generator implementation _as it exists on disk when you open the solution_ for the duration of that session. This means in exchange for the simplicity of defining a source generator in the same solution, you're expected to avoid certain behaviors that lead to problems:

* The solution should be built at least once before opening the solution (or if you build inside the IDE, restart after the first time you build the source generator project)
* Avoid deleting the bin/obj directory of the source generator (or if you do, restart at the previous step)
* If you make changes to the source generator implementation, you'll have to build it and then restart the IDE for IntelliSense to pick up the changes

**Comment by Sergio0694** [comment]
Thank you for all the additional info Sam, this helps a lot! 😄
I was able to be much more productive now that I know what was causing things to stop working.
I would say though that especially the last bullet point is particularly frustrating, having to restart VS every single time I change one line makes developing a source generator quite the challenge. It'd be nice if Roslyn could automatically reload it instead.

> "except for cases where the source generator implementation has been changed since last build, in which case it would update as soon as the project is next built"

I'm probably missing context here, but if the generator is referenced locally from another project in the same solution, couldn't Roslyn see that it has been modified after the last build, and rebuild it first and reload it before starting building the project?

As in, what I mean is:
- I have a generator G
- I have a proper P referencing G as an analyzer
- I make some changes to files in G
- I then build/rebuild/click play on P

I would expect VS to see G has been changed since the last build and that P is referencing G, and therefore first rebuild G, reload it, generate the files with it and _then_ start building P. Similarly, after building P in this scenario, I'd also expect the list of generated files from the solution explorer to correctly be updated and show the new output from G under P's tree.

Would this be possible? 🤔
